Sprint 7 solution in file manager

glkurkin opened the pull request.

glkurkin committed:
Объединение папок test и tests

glkurkin committed:
Реализован менеджер задач с поддержкой сохранения в файл и связанные изменения

glkurkin committed:
Реализован менеджер задач с поддержкой сохранения в файл и связанные изменения

glkurkin committed:
Update TaskType.java

Не проходил CheckStyle, приписывал, что неверное имя пакета, хотя его не должно быть

sergeymamonov commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,99 @@
+import java.io.*;
```
Не используйте импорты через *.

sergeymamonov commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,99 @@
+                writer.write(taskToString(task));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Ошибка сохранения в файл", e);
```
Здесь нужно кинуть не RuntimeException (это общее исключение), а свое исключение.

sergeymamonov commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,99 @@
+
+    protected void save() {
+        try (Writer writer = new FileWriter(file)) {
+            writer.write("id,type,name,status,description,epic\n");
```
Литералы лучше выносить из кода. Сделайте константу с информативным названием.

sergeymamonov commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,99 @@
+        );
+    }
+
+    public static FileBackedTaskManager loadFromFile(File file) {
```
Не вижу, чтобы этот метод где-то использовался.

sergeymamonov commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,99 @@
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Ошибка загрузки из файла", e);
```
Свое исключение.

sergeymamonov commented on src/InMemoryTaskManager.java:
```diff
@@ -4,9 +4,9 @@
-    private Map<Integer, Task> tasks;
-    private Map<Integer, Epic> epics;
-    private Map<Integer, Subtask> subtasks;
+    protected Map<Integer, Task> tasks;
```
Лучше оставить приватный модификатор и обращаться через геттеры и сеттеры.

sergeymamonov commented on src/InMemoryTaskManager.java:
```diff
@@ -4,9 +4,9 @@
+    protected Map<Integer, Subtask> subtasks;
     private int id;
     private HistoryManager historyManager;
 
```
В методах встречается строка int id = this.id++;. Так делать не очень хорошо, потому что происходит перекрытие (затенение) поля класса локальной переменной. Как вариант изменить имя одной из переменных.

sergeymamonov commented on test/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,4 @@
+import static org.junit.jupiter.api.Assertions.*;
+class FileBackedTaskManagerTest {
+  
```
Лучше добавить тесты и не использовать импорт через *. Для этого отключить wildcard import https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcards-for-class

sergeymamonov reviewed (changes requested):
.DS_Store лучше добавить в гитигнор, чтобы он не попадал в коммиты.
Классов становится все больше, поэтому лучше раскладывать их по пакетам, чтобы было легче оринтироваться в проекте.
Файл test.txt - если это генерируемый файл в ходе выполнения тестов, то лучше его тоже не включать в коммиты и прописать в гитигноре.

glkurkin committed:
Update FileBackedTaskManager.java

Исправлены замечания с импортом, созданы собственные исключения, вынесены литералы и переименована переменная с id на taskId

glkurkin committed:
Add files via upload

glkurkin committed:
Update FileBackedTaskManagerTest.java

glkurkin committed:
Update FileSaveException.java
